inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/dataprovider/generic/DataProviderGenerator.java

```diff
@@ -0,0 +1,213 @@
+import java.util.concurrent.ExecutionException;
+
+@Component
+@Slf4j
+public class DataProviderGenerator {
+
+    /**
+     * Guava seems to not allow null keys.
+     * Therefore we just use this object as replacement for the bootstrap loader.
+     */
+    private static final ClassLoader BOOTSTRAP_LOADER_MARKER = new URLClassLoader(new URL[]{});
+
+    private static final String GENERIC_PROVIDER_STRUCTURAL_ID = "genericDataProvider";
+
+    private static final String METHOD_ARGS = "$1";
+    private static final String THIZ = "$2";
+    private static final String RETURN_VALUE = "$3";
+    private static final String THROWN = "$4";
+    private static final String ADDITIONAL_ARGS = "$5";
+
+    public static final String THIZ_VARIABLE = "thiz";
+    public static final String ARGS_VARIABLE = "args";
+    public static final String ARG_VARIABLE_PREFIX = "arg";
+    public static final String THROWN_VARIABLE = "thrown";
+    public static final String RETURN_VALUE_VARIABLE = "returnValue";
+
+    @Autowired
+    private ClassInjector classInjector;
+
+    private LoadingCache<ClassLoader, Cache<ResolvedGenericDataProviderConfig, InjectedClass<? extends IGenericDataProvider>>> providersCache
+            = CacheBuilder.newBuilder().weakKeys().build(
+            new CacheLoader<ClassLoader, Cache<ResolvedGenericDataProviderConfig, InjectedClass<? extends IGenericDataProvider>>>() {
+                @Override
+                public Cache<ResolvedGenericDataProviderConfig, InjectedClass<? extends IGenericDataProvider>> load(ClassLoader key) {
+                    return CacheBuilder.newBuilder().weakValues().build();
+                }
+            });
+
+    /**
+     * Provides an executable {@link IGenericDataProvider} based on the given configuration.
+     * The provider is either dynamically compiled and injected or a cached provider is used.
+     *
+     * @param providerConfig       the configuration of the generic data provider to use
+     * @param classToUseProviderOn the context in which the provider will be actively. The provider will be injected into the classloader of this class.
+     * @return the generated provider
+     * @throws ExecutionException
+     */
+    @SuppressWarnings("unchecked")
+    public InjectedClass<? extends IGenericDataProvider> getOrGenerateDataProvider(ResolvedGenericDataProviderConfig providerConfig, Class<?> classToUseProviderOn) {
+        ClassLoader loader = Optional.ofNullable(classToUseProviderOn.getClassLoader()).orElse(BOOTSTRAP_LOADER_MARKER);
+        providersCache.cleanUp();
+        Cache<ResolvedGenericDataProviderConfig, InjectedClass<? extends IGenericDataProvider>> clCache;
+        try {
+            clCache = providersCache.get(loader);
+            clCache.cleanUp(); //cleanup to make sure unused InjectedClasses are released
+            try {
+                return clCache.get(providerConfig, () ->
+                        (InjectedClass<? extends IGenericDataProvider>)
```

Comment:
why is this cast introduced?

Reply:
I think this cast was tehre previously..
The ClassInjector of course does not know what stuff you inject, therefore you have to cast it manually.